### ISSUE TITLE
LibGfx: Google Doodle Issue #3

### ISSUE THREAD

**Issue by Lubrsi** [SerenityOS/serenity#4260]
The [current Google Doodle](https://www.google.co.uk/logos/doodles/2020/st-andrews-day-2020-6753651837108633.3-l.png) does not successfully decode:
![vmplayer_SXUpmEFzyH](https://user-images.githubusercontent.com/25595356/100557907-cfa9c580-32a3-11eb-8dc0-fafe7724a0a2.png)

Opening the image in a separate tab causes an assertion failure:
```
USERSPACE(60) ASSERTION FAILED: bitmap
../Libraries/LibWeb/Loader/FrameLoader.cpp:95
[Browser(60:60)]: Terminating Browser(60) due to signal 6
[FinalizerTask(3:3)]: CPU[0] NP(error) fault at invalid address V0x022025ec
[FinalizerTask(3:3)]: copy_from_user(0xc094a92c, 0x022025ec, 4) failed at V0x022025ec
[FinalizerTask(3:3)]: 0xdeadc0de  Kernel::Processor::switch_context(Kernel::Thread*&, Kernel::Thread*&) +181
0xdeadc0de  Kernel::Scheduler::context_switch(Kernel::Thread*) +181
0xdeadc0de  Kernel::Scheduler::pick_next() +2280
0xdeadc0de  Kernel::Processor::check_invoke_scheduler() +125
0xdeadc0de  Kernel::Thread::die_if_needed() +157
0xdeadc0de  syscall_handler +1055
0xdeadc0de  syscall_asm_entry +49
0x0824cc1c  raise +22
0x080480bd   +13
0x08248235  tolower +0
0x080d33f9  Web::FrameLoader::create_document_from_mime_type(AK::ByteBuffer const&, AK::URL const&, AK::String const&, AK::String const&) +391
0x080d5167  Web::FrameLoader::resource_did_load() +399
0x080d74b7  Web::Resource::for_each_client(AK::Function<void (Web::ResourceClient&)>) +309
0x080d7aa1  Web::Resource::did_load(AK::Badge<Web::ResourceLoader>, AK::ByteBuffer const&, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&) +1191
0x080d85b2  AK::Function<void (AK::ByteBuffer const&, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&)>::CallableWrapper<Web::ResourceLoader::load_resource(Web::Resource::Type, Web::LoadRequest const&)::{lambda(auto:1&, auto:2&)#1}>::call(AK::ByteBuffer const&, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&) const +46
0x080d8deb  AK::Function<void (bool, AK::ByteBuffer const&, AK::RefPtr<AK::SharedBuffer, AK::RefPtrTraits<AK::SharedBuffer> >, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&, AK::Optional<unsigned int>)>::CallableWrapper<Web::ResourceLoader::load(Web::LoadRequest const&, AK::Function<void (AK::ByteBuffer const&, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&)>, AK::Function<void (AK::String const&)>)::{lambda(bool, AK::ByteBuffer const&, auto:1, auto:2&, auto:3)#4}>::call(bool, AK::ByteBuffer const&, AK::RefPtr<AK::SharedBuffer, AK::RefPtrTraits<AK::SharedBuffer> >, AK::HashMap<AK::String, AK::String, AK::CaseInsensitiveStringTraits> const&, AK::Optional<unsigned int>) const +215
0x0813185a  Protocol::Download::did_finish(AK::Badge<Protocol::Client>, bool, AK::Optional<unsigned int>, unsigned int, int, IPC::Dictionary const&) +742
0x0812df7e  Protocol::Client::handle(Messages::ProtocolClient::DownloadFinished const&) +254
0x0812e9db  ProtocolClientEndpoint::handle(IPC::Message const&) +53
0x0813137d  IPC::Connection<ProtocolClientEndpoint, ProtocolServerEndpoint>::handle_messages() +123
0x08245270  Core::Object::dispatch_event(Core::Event&, Core::Object*) +74
0x0823e23d  Core::EventLoop::pump(Core::EventLoop::WaitMode) +397
0x0823e4b3  Core::EventLoop::exec() +43
0x08048cda  main +2266
0x0805156e  _start +94
```

**Comment by linusg** [comment]
`PNG_DEBUG` output:

```
Chunk type: 'IHDR', size: 13, crc: 676292fa
PNG: 490x196 (8 bpp)
     Color type: 3
Compress Method: 0
  Filter Method: 0
 Interlace type: 0
Chunk type: 'PLTE', size: 765, crc: 351f053
Chunk type: 'tRNS', size: 255, crc: 64cc411
Chunk type: 'IDAT', size: 55011, crc: 34f37573
Chunk type: 'IEND', size: 0, crc: ae426082
```

It fails here:

https://github.com/SerenityOS/serenity/blob/34c5478f3133a6d235e91c8004b102f369e842b5/Libraries%2FLibGfx%2FPNGLoader.cpp#L753-L754

With:

```
context.color_type = 3
context.palette_data.size() = 255
(1u << context.bit_depth) = 256
```

**Comment by alimpfard** [comment]
Section 4.1.2 in https://www.w3.org/TR/REC-png-961001
> For color type 3 (indexed color), the PLTE chunk is required. The first entry in PLTE is referenced by pixel value 0, the second by pixel value 1, etc. The number of palette entries must not exceed the range that can be represented in the image bit depth (for example, 2^4 = 16 for a bit depth of 4). **It is permissible to have fewer entries than the bit depth would allow. In that case, any out-of-range pixel value found in the image data is an error.**

**Comment by linusg** [comment]
Fixed by 92fe63fbecedb1266e472861aba9afed7d8f87f9.

![image](https://user-images.githubusercontent.com/19366641/103767737-7448c100-5021-11eb-8545-607a07b1f921.png)

(this is the image mentioned above)
